container-engine-lib/lib/backend_impls/docker/docker_manager/docker_manager.go

```diff
@@ -1292,6 +1305,119 @@ func (manager *DockerManager) FetchImage(ctx context.Context, image string, down
 	return pulledFromRemote, imageArchitecture, nil
 }
 
+func (manager *DockerManager) BuildImage(ctx context.Context, imageName string, imageBuildSpec *image_build_spec.ImageBuildSpec) (string, error) {
+	contextDirPath := imageBuildSpec.GetContextDirPath()
+	containerImageFileTarReader, err := getBuildContextReader(contextDirPath)
+	if err != nil {
+		return "", stacktrace.Propagate(err, "An error occurred retrieving the build context for '%v' at context directory path: %v", imageName, contextDirPath)
+	}
+
+	// Before instructing docker client to execute an image build, we need to create a connection to buildkit
+	// buildkit is the daemon process that executes build workloads: https://docs.docker.com/build/architecture/#buildkit
+
+	// Setup session to buildkit (eg. https://github.com/hashicorp/waypoint/pull/1937)
+	uuidStr, err := uuid_generator.GenerateUUIDString()
+	if err != nil {
+		return "", stacktrace.Propagate(err, "An error occurred generating a UUID to give the Docker Buildkit session")
+	}
+	sessionName := fmt.Sprintf("kurtosis-%s", uuidStr)
+
+	// Generate a new session every time because per https://github.com/moby/buildkit/issues/1432 sharing sessions is an optimization
+	// Don't bother reusing sessions so that we don't hit bugs
+	buildkitSession, err := bksession.NewSession(ctx, sessionName, buildkitSessionSharedKey)
+	if err != nil {
+		return "", stacktrace.Propagate(err, "An error generating a Docker Buildkit session with sessionName: %v", sessionName)
+	}
+	dialSessionFunc := func(ctx context.Context, proto string, meta map[string][]string) (net.Conn, error) {
+		return manager.dockerClientNoTimeout.DialHijack(ctx, "/session", proto, meta)
+	}
+
+	// Activate the session
+	go buildkitSession.Run(ctx, dialSessionFunc) // nolint
```

Comment:
the error is not handled here, is it because the `ImageBuild` call will fail below? Could you at least log the error?